servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSImpl.java

```diff
@@ -1856,6 +1898,79 @@ void validateDomainValues(Domain domain) {
+        }
+        if (!domainMetaStore.isValidAzureSubscription(domain.getName(), domain.getAzureSubscription())) {
+            throw ZMSUtils.requestError("invalid azure subscription for domain", caller);
+        }
+        if (!domainMetaStore.isValidProductId(domain.getName(), domain.getYpmId())) {
+            throw ZMSUtils.requestError("invalid product id for domain", caller);
+        }
+    }
+
+    BitSet validateDomainSystemMetaStoreValues(Domain domain, DomainMeta meta, final String attributeName) {
+
+        final String caller = "validateDomainSystemMetaStoreValues";
+
+        // we're going to check the meta values for our new domain
+        // requests against our meta store.
+
+        BitSet changedAttrs = new BitSet();
+        switch (attributeName) {
+            case ZMSConsts.SYSTEM_META_ACCOUNT:
+                if (ZMSUtils.metaValueChanged(domain.getAccount(), meta.getAccount())) {
+                    if (!domainMetaStore.isValidAWSAccount(domain.getName(), meta.getAccount())) {
+                        throw ZMSUtils.requestError("invalid aws account for domain", caller);
+                    }
+                    changedAttrs.set(DomainMetaStore.META_ATTR_AWS_ACCOUNT);
+                }
+                break;
+            case ZMSConsts.SYSTEM_META_AZURE_SUBSCRIPTION:
+                if (ZMSUtils.metaValueChanged(domain.getAzureSubscription(), meta.getAzureSubscription())) {
+                    if (!domainMetaStore.isValidAzureSubscription(domain.getName(), meta.getAzureSubscription())) {
+                        throw ZMSUtils.requestError("invalid azure subscription for domain", caller);
+                    }
+                    changedAttrs.set(DomainMetaStore.META_ATTR_AZURE_SUBSCRIPTION);
+                }
+                break;
+            case ZMSConsts.SYSTEM_META_PRODUCT_ID:
+                if (ZMSUtils.metaValueChanged(domain.getYpmId(), meta.getYpmId())) {
+                    if (!domainMetaStore.isValidProductId(domain.getName(), meta.getYpmId())) {
+                        throw ZMSUtils.requestError("invalid product id for domain", caller);
+                    }
+                    changedAttrs.set(DomainMetaStore.META_ATTR_PRODUCT_ID);
+                }
+                break;
+        }
+
+        return changedAttrs;
+    }
+
+    BitSet validateDomainRegularMetaStoreValues(Domain domain, DomainMeta meta) {
+
+        final String caller = "validateDomainRegularMetaStoreValues";
+
+        // we're going to check the meta values for our new domain
+        // requests against our meta store - for now we only have
+        // business service as a regular domain meta attribute
+
+        BitSet changedAttrs = new BitSet();
+        if (ZMSUtils.metaValueChanged(domain.getBusinessService(), meta.getBusinessService())) {
+            if (!domainMetaStore.isValidBusinessService(domain.getName(), meta.getBusinessService())) {
```

Comment:
This question applies to all meta attributes.

Lets say domain A has business service X, and its been changed to value Y, whats the purpose of passing domainName to domainMetaStore.isValidBusinessService()
How does a real life validation look like? e.g Business service not associated with any other domain?

Reply:
it's completely up to the implementation. For example, I have an external service that already maintains the mapping of athenz domains and business service. Since the domain admin has the rights to set the business service I don't want them to set to some random value. So the implementation will call isValidBusinessService for domain A and value Y. If the external store says it must be service X, then the request will be rejected so that the domain admin can't set invalid value Y in Athenz.

Reply:
The other use case would be that it's just completely ignored. I only want to check if the value is an actual business service or not. In that case the implementation will just ignore the domainName and will only verify that businessService value is valid or not.